homeassistant/components/zha/climate.py

```diff
@@ -0,0 +1,550 @@
+        if self._thrm.occupancy is not None:
+            data[ATTR_OCCUPANCY] = self._thrm.occupancy
+        if self._thrm.occupied_cooling_setpoint is not None:
+            data[ATTR_OCCP_COOL_SETPT] = self._thrm.occupied_cooling_setpoint
+        if self._thrm.occupied_heating_setpoint is not None:
+            data[ATTR_OCCP_HEAT_SETPT] = self._thrm.occupied_heating_setpoint
+
+        unoccupied_cooling_setpoint = self._thrm.unoccupied_cooling_setpoint
+        if unoccupied_cooling_setpoint is not None:
+            data[ATTR_UNOCCP_HEAT_SETPT] = unoccupied_cooling_setpoint
+
+        unoccupied_heating_setpoint = self._thrm.unoccupied_heating_setpoint
+        if unoccupied_heating_setpoint is not None:
+            data[ATTR_UNOCCP_COOL_SETPT] = unoccupied_heating_setpoint
+        return data
+
+    @property
+    def fan_mode(self) -> Optional[str]:
+        """Return current FAN mode."""
+        if self._thrm.running_state is None:
+            return FAN_AUTO
+
+        if self._thrm.running_state & (
+            RunningState.FAN | RunningState.FAN_STAGE_2 | RunningState.FAN_STAGE_3
+        ):
+            return FAN_ON
+        return FAN_AUTO
+
+    @property
+    def fan_modes(self) -> Optional[List[str]]:
+        """Return supported FAN modes."""
+        if not self._fan:
+            return None
+        return [FAN_AUTO, FAN_ON]
+
+    @property
+    def hvac_action(self) -> Optional[str]:
+        """Return the current HVAC action."""
+        if (
+            self._thrm.pi_heating_demand is None
+            and self._thrm.pi_cooling_demand is None
+        ):
+            self.debug("Running mode: %s", self._thrm.running_mode)
+            self.debug("Running state: %s", self._thrm.running_state)
+            running_state = self._thrm.running_state
+            if running_state is None:
+                return None
+            if running_state & (RunningState.HEAT | RunningState.HEAT_STAGE_2):
+                return CURRENT_HVAC_HEAT
+            if running_state & (RunningState.COOL | RunningState.COOL_STAGE_2):
+                return CURRENT_HVAC_COOL
+            if running_state & (
+                RunningState.FAN | RunningState.FAN_STAGE_2 | RunningState.FAN_STAGE_3
+            ):
+                return CURRENT_HVAC_FAN
+        else:
+            heating_demand = self._thrm.pi_heating_demand
+            if heating_demand is not None and heating_demand > 0:
+                return CURRENT_HVAC_HEAT
+            cooling_demand = self._thrm.pi_cooling_demand
+            if cooling_demand is not None and cooling_demand > 0:
+                return CURRENT_HVAC_COOL
+        if self.hvac_mode != HVAC_MODE_OFF:
+            return CURRENT_HVAC_IDLE
+        return CURRENT_HVAC_OFF
+
+    @property
+    def hvac_mode(self) -> Optional[str]:
+        """Return HVAC operation mode."""
+        try:
+            return SYSTEM_MODE_2_HVAC[self._thrm.system_mode]
+        except KeyError:
+            self.error(
```

Comment:
If this is a log message, we don't log inside entity state properties. It's a side effect.

Reply:
what are my options to collect info about thermostats which report `system_mode` not defines in the specs? 
Should I just let it throw?

Reply:
We usually use a callback function to update instance attributes and handle all side effects there, if needed. Then in the property just return the instance attribute.

Reply:
ok